src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -293,23 +297,35 @@ private void OnWriteCompleted(WriteContext writeContext)
             {
                 PoolWriteContext(writeContext);
 
-                // _numBytesPreCompleted can temporarily go negative in the event there are
-                // completed writes that we haven't triggered callbacks for yet.
-                _numBytesPreCompleted -= bytesWritten;
-
-                // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
-                // This allows large writes to complete once they've actually finished.
-                var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
-                while (_tasksPending.Count > 0 &&
-                       (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                if (_lastWriteError == null)
                 {
-                    var tcs = _tasksPending.Dequeue();
-                    var bytesToWrite = (int)tcs.Task.AsyncState;
+                    // _numBytesPreCompleted can temporarily go negative in the event there are
+                    // completed writes that we haven't triggered callbacks for yet.
+                    _numBytesPreCompleted -= bytesWritten;
+
+                    // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
+                    // This allows large writes to complete once they've actually finished.
+                    var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
+                    while (_tasksPending.Count > 0 &&
+                           (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                    {
+                        var tcs = _tasksPending.Dequeue();
+                        var bytesToWrite = (int)tcs.Task.AsyncState;
 
-                    _numBytesPreCompleted += bytesToWrite;
-                    bytesLeftToBuffer -= bytesToWrite;
+                        _numBytesPreCompleted += bytesToWrite;
+                        bytesLeftToBuffer -= bytesToWrite;
 
-                    _tasksCompleted.Enqueue(tcs);
+                        _tasksCompleted.Enqueue(tcs);
+                    }
+                }
+                else
+                {
+                    // Error all waiting tasks
+                    while (_tasksPending.Count > 0)
```

Comment:
> so one of us is going to have to rebase depending on the order these get in (and we know who controls that :stuck_out_tongue:

:laughing:

> Another consideration is how this is going to interact with the per-request IgnoreWriteExceptions property we are likely to add.

Isn't too hard; though would be an issue if someone was streaming something large but the client had disconnected as if they weren't listening for it they'd just keep writing to null?